File: mqttclients/dotnet/MQTTnet.Client.Extensions/Utf8JsonSerializer.cs
using System.Text.Json;

namespace MQTTnet.Client.Extensions;
public class Utf8JsonSerializer : IMessageSerializer
{
    public string ContentType => "application/json";

    public T FromBytes<T>(byte[] payload)
    {
        Utf8JsonReader reader = new Utf8JsonReader(payload);
        return JsonSerializer.Deserialize<T>(ref reader)!;
    }

    public byte[] ToBytes<T>(T payload) => JsonSerializer.SerializeToUtf8Bytes(payload);
}

File: mqttclients/ts/mqttjsClientExtensions/src/utf8JsonSerializer.ts
import { IMessageSerializer } from ".";

export class Utf8JsonSerializer<T> implements IMessageSerializer {
    public contentType = 'application/json';

    public fromBytes<T>(payload: Buffer): T {
        return JSON.parse(payload.toString('utf8')) as T;
    }

    public toBytes<T>(payload: T): Buffer {
        return Buffer.from(JSON.stringify(payload), 'utf8');
    }
}
